sktime/transformations/panel/window_summarizer.py

```diff
@@ -0,0 +1,268 @@
+
+from sktime.transformations.base import _PanelToTabularTransformer
+
+# List of native pandas rolling window function.
+# danbartl: investivate different engines for pandas
+pd_rolling = [
+    "sum",
+    "mean",
+    "median",
+    "std",
+    "var",
+    "kurt",
+    "min",
+    "max",
+    "corr",
+    "cov",
+    "skew",
+    "sem",
+]
+
+
+def _window_feature(
+    Z,
+    name=None,
+    win_summarizer=None,
+    window=None,
+):
+    """Compute window features and lag.
+
+    Apply functions passed over a certain window
+    of past observations, e.g. the mean of a window of length 7 days, lagged by 14 days.
+
+    y: either pandas.core.groupby.generic.SeriesGroupBy
+        Object create by grouping across groupBy columns.
+    name : str, base string of the derived features
+           final name will also contain
+           window shift and window length.
+    win_summarizer: either str corresponding to native
+    implemented pandas function, currently
+            * "sum",
+            * "mean",
+            * "median",
+            * "std",
+            * "var",
+            * "kurt",
+            * "min",
+            * "max",
+            * "corr",
+            * "cov",
+            * "skew",
+            * "sem"
+         or function definition. See for the native functions also
+         https://pandas.pydata.org/docs/reference/window.html.
+    window: list of integers
+        Contains values for window shift and window length.
+    """
+    win = window[0]
+    shift = window[1] + 1
+
+    if win_summarizer in pd_rolling:
+        if isinstance(Z.index, pd.MultiIndex):
+            feat = getattr(Z.shift(shift).rolling(win), win_summarizer)()
+        else:
+            feat = pd.DataFrame(Z).apply(
+                lambda x: getattr(x.shift(shift).rolling(win), win_summarizer)()
+            )
+    else:
+        feat = Z.shift(shift)
+        if isinstance(Z.index, pd.MultiIndex) and callable(win_summarizer):
+            feat = feat.rolling(win).apply(win_summarizer, raw=True)
+        elif not isinstance(Z.index, pd.MultiIndex) and callable(win_summarizer):
+            feat = feat.apply(lambda x: x.rolling(win).apply(win_summarizer, raw=True))
+
+    if isinstance(feat, pd.Series):
+        feat = pd.DataFrame(feat)
+
+    feat.rename(
+        columns={
+            feat.columns[0]: name + "_" + "_".join([str(item) for item in window])
+        },
+        inplace=True,
+    )
+
+    return feat
+
+
+class _LaggedWindowExtractor(_PanelToTabularTransformer):
+    """Base adapter class for transformations.
+
+    The LaggedWindowSummarizer transforms input series to features
+    based on a provided dictionary of window summarizer, window shifts
+    and window lengths.
+    """
+
+    _tags = {
+        # todo: what is the scitype of X: Series, or Panel
+        # "scitype:transform-input": "Panel",
+        # todo: what scitype is returned: Primitives, Series, Panel
+        "scitype:transform-output": "Panel",
```

Comment:
should this not be `Series`? Since we transform the series into a series that has the lagged summaries?

Reply:
Not sure, the transform can take Series and Pandas DataFrame as inputs and will return a Pandas Dataframe (typically one Series / column gets transformed to multiple columns). Panel is not supported due to the changes required that I factored out (part of hierarchical PR). So what should I choose here?

Reply:
I think it should be `Series`, that´s for all univariate series formats (`pd.Series` and `pd.DataFrame`) - let´s see what happens...

Reply:
ok just tried:

        "scitype:transform-input": "Series",
        "scitype:transform-output": "Series",

This works in principle , but then the following happens:

-  Input pd.DataFrame with one argument in kwargs -> pd.DataFrame Output with one column
- Input pd.DataFrame with multiple arguments in kwargs -> pd.DataFrame Output with multiple column
- Input pd.Series with one argument in kwargs -> **pd.Series Output** with one column
- Input pd.Series with multiple arguments in kwargs -> pd.DataFrame Output with multiple column


I personally would prefer to have always the same output (pd.DataFrame), especially since it now depends on kwargs argument. How can I make sure this happens for case 3? Currently the output gets automatically converted.

Thanks!

Reply:
> I personally would prefer to have always the same output (pd.DataFrame), especially since it now depends on kwargs argument. What do you think?

Currently, what you describe is *default behaviour* of the interface, of all forecasters.

It has been discussed internally and decided that this is a bit confusing, and the output should always be `pd.DataFrame`. The change will likely happen in 0.11.0, through changes in the `BaseForecaster`. I.e., it will automatically be applied to all forecasters and yours through the `BaseForecaster` change.

Reply:
current behaviour is "input type = output type (as long as it is possible)", it will be replaced with "constant output type"